src/main/java/org/spongepowered/common/launch/transformer/tracker/TrackerClassTransformer.java

```diff
@@ -0,0 +1,210 @@
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.launch.transformer.tracker;
+
+import static org.spongepowered.asm.lib.Opcodes.ACC_PRIVATE;
+import static org.spongepowered.asm.lib.Opcodes.ACC_STATIC;
+import static org.spongepowered.asm.lib.Opcodes.ALOAD;
+import static org.spongepowered.asm.lib.Opcodes.ARETURN;
+import static org.spongepowered.asm.lib.Opcodes.ASM5;
+import static org.spongepowered.asm.lib.Opcodes.CHECKCAST;
+import static org.spongepowered.asm.lib.Opcodes.DLOAD;
+import static org.spongepowered.asm.lib.Opcodes.DRETURN;
+import static org.spongepowered.asm.lib.Opcodes.FLOAD;
+import static org.spongepowered.asm.lib.Opcodes.FRETURN;
+import static org.spongepowered.asm.lib.Opcodes.F_SAME;
+import static org.spongepowered.asm.lib.Opcodes.IFEQ;
+import static org.spongepowered.asm.lib.Opcodes.ILOAD;
+import static org.spongepowered.asm.lib.Opcodes.INSTANCEOF;
+import static org.spongepowered.asm.lib.Opcodes.INVOKEINTERFACE;
+import static org.spongepowered.asm.lib.Opcodes.INVOKESTATIC;
+import static org.spongepowered.asm.lib.Opcodes.INVOKEVIRTUAL;
+import static org.spongepowered.asm.lib.Opcodes.IRETURN;
+import static org.spongepowered.asm.lib.Opcodes.LLOAD;
+import static org.spongepowered.asm.lib.Opcodes.LRETURN;
+import static org.spongepowered.asm.lib.Opcodes.RETURN;
+
+import net.minecraft.launchwrapper.IClassTransformer;
+import org.spongepowered.asm.lib.ClassReader;
+import org.spongepowered.asm.lib.ClassWriter;
+import org.spongepowered.asm.lib.Label;
+import org.spongepowered.asm.lib.MethodVisitor;
+import org.spongepowered.asm.lib.Type;
+import org.spongepowered.asm.lib.tree.AbstractInsnNode;
+import org.spongepowered.asm.lib.tree.ClassNode;
+import org.spongepowered.asm.lib.tree.MethodInsnNode;
+import org.spongepowered.asm.lib.tree.MethodNode;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.ListIterator;
+import java.util.Map;
+import java.util.Set;
+
+public class TrackerClassTransformer implements IClassTransformer {
+
+    @Override
+    public byte[] transform(String name, String transformedName, byte[] basicClass) {
```

Comment:
> When could one of these parameters be null?

`basicClass` can be `null` when the class is non-existent or is expected to be generated at runtime, aka *runtime-synthetic* class. If your transformer doesn't synthesize the class then it should return `null` per the transformer contract, throwing a `NullPointerException` when trying to read the class bytecode which doesn't exist is not an acceptable behaviour for a transformer because it leads to unexpected `ClassNotFoundError`s which are actually because the transformer which *should* generate the class never gets a chance to. For example Mixin generates different classes at runtime to support target-specialised inner classes, and `Args` subclasses for `@ModifyArgs`.

> It doesn't seem to be handled in SpongeSuperclassTransformer as well:

Then that's a mistake and should be rectified, and the person who wrote it duly chastised.